### PR TITLE
fix(ldap-auth) fix sAMAccountName and UPN attribute consideration

### PR DESCRIPTION
**Pull request by littlechicks**
Original ldap-auth plugins only consider CommonName and UID attributes.
In fact the plugins is always use a DistinguisedName for the authenticate.
However, the use of attributes like sAMAccount or UPN don't work with a DN based authentification.
The appropriate way to use sAMAccountName and UPN is to allow LDAP authentification with
an "who" user like : sAMAccountName@DOMAIN or DOMAIN\sAMAccountName.
But the clean way is to use the first suggestion.

https://github.com/Kong/kong/issues/5121

### Summary

Attributes in LDAP can be used to bind and user. In case of LDAP Auth plugins of Kong, only the CN is working. Other attributes like sAMAccountName and UPN cannot be used regardless the code that consider only a binding based on DN syntax.

### Full changelog

* Add consideration to differents attributes for binding

### Issues resolved

Fix https://github.com/Kong/kong/issues/5121
https://github.com/Kong/kong/issues/3705

### CODE REVIEW

**Comment by kikito** [comment]
Hi @littlechicks , 

Thanks for cleaning up the PR. I have two comments:

* Are these changes part of some LDAP specification?
* Would you be able to include tests for this change in your PR as well?

**Comment by littlechicks** [comment]
Hi @kikito ,

To answer your comments: 

- Yes and no. The changes are part of Active Directory specification.
In fact, in LDAP Specification (https://www.ietf.org/rfc/rfc4514.txt) only CN and UID are valid attribute for binding. Therefore, they can be used within a Distinguished Name.

But let's clear this point.

In the `access.lua` file we call the function `ldap.bindrequest` which takes a username and password as arguments.
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/access.lua#L95

Now in the `ldap.lua` we have the function _M.bind_request let's focus on theses lines:
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/ldap.lua#L67-L70

- The first line format 0x80 type code which a "simple bind" type with the password.

- On the second line we create the LDAPMessage with the protocol version (here "3" for LDAPv3) then the username and finally the block created on the first line.

- Finally we create de full message by adding the message "ID" with the bindRequest code which is 0, and format the whole message

So finally we have something like that for the payload:

payload
 

I have tested a simple code for binding a user on an Active Directory LDAP using _UserPrincipalName_ attribute (_username@domain_) and I've checked the payload in Wireshark, **without SSL**.

It's look like this:

capture

So, we see that it is exactly the same payload than the LDAP Specification and the KONG `ldap.lua`file.
Therefore, we can use the same code without changing anything to the bind request function.
The only thing to add is what I've done in my PR. (_Add UPN and sAMAccountName as argument for the bind request_)


- For the tests, I will try to do some code but I am very not a expert of lua...But i'll try.

**Comment by littlechicks** [comment]
Hi @kikito ,

To answer your comments: 

- Yes and no. The changes are part of Active Directory specification.
In fact, in LDAP Specification (https://www.ietf.org/rfc/rfc4514.txt) only CN and UID are valid attribute for binding. Therefore, they can be used within a Distinguished Name.

But let's clear this point.

In the `access.lua` file we call the function `ldap.bindrequest` which takes a username and password as arguments.
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/access.lua#L95

Now in the `ldap.lua` we have the function _M.bind_request let's focus on theses lines:
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/ldap.lua#L67-L70

- The first line format 0x80 type code which a "simple bind" type with the password.

- On the second line we create the LDAPMessage with the protocol version (here "3" for LDAPv3) then the username and finally the block created on the first line.

- Finally we create de full message by adding the message "ID" with the bindRequest code which is 0, and format the whole message

So finally we have something like that for the payload:

payload
 

I have tested a simple code for binding a user on an Active Directory LDAP using _UserPrincipalName_ attribute (_username@domain_) and I've checked the payload in Wireshark, **without SSL**.

It's look like this:


payload


So, we see that it is exactly the same payload than the LDAP Specification and the KONG `ldap.lua`file.
Therefore, we can use the same code without changing anything to the bind request function.
The only thing to add is what I've done in my PR. (_Add UPN and sAMAccountName as argument for the bind request_)


- For the tests, I will try to do some code but I am very not a expert of lua...But i'll try.

**Comment by littlechicks** [comment]
Hi @kikito ,

To answer your comments: 

- Yes and no. The changes are part of Active Directory specification.
In fact, in LDAP Specification (https://www.ietf.org/rfc/rfc4514.txt) only CN and UID are valid attribute for binding. Therefore, they can be used within a Distinguished Name.

But let's clear this point.

In the `access.lua` file we call the function `ldap.bindrequest` which takes a username and password as arguments.
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/access.lua#L95

Now in the `ldap.lua` we have the function _M.bind_request let's focus on theses lines:
https://github.com/Kong/kong/blob/0c71caac07eade4a567a1ef1e1992f07d5a58c8c/kong/plugins/ldap-auth/ldap.lua#L67-L70

- The first line format 0x80 type code which a "simple bind" type with the password.

- On the second line we create the LDAPMessage with the protocol version (here "3" for LDAPv3) then the username and finally the block created on the first line.

- Finally we create de full message by adding the message "ID" with the bindRequest code which is 0, and format the whole message

So finally we have something like that for the payload:

payload
 

I have tested a simple code for binding a user on an Active Directory LDAP using _UserPrincipalName_ attribute (_username@domain_) and I've checked the payload in Wireshark, **without SSL**.

It's look like this:


capture


So, we see that it is exactly the same payload than the LDAP Specification and the KONG `ldap.lua`file.
Therefore, we can use the same code without changing anything to the bind request function.
The only thing to add is what I've done in my PR. (_Add UPN and sAMAccountName as argument for the bind request_)


- For the tests, I will try to do some code but I am very not a expert of lua...But i'll try.

**Comment by littlechicks** [comment]
So........... @kikito 

I cannot include tests...because Kong Spec for ldap plugins use `ec2-54-172-82-117.compute-1.amazonaws.com`
And this LDAP Backend is an OpenLDAP, and not Active Directory.
So I cannot test an authentication with UserPrincipalName or SAMAccoutName which are specific to MS-Ad.

See bellow the error.

test_ldap

`The successful binding is when I use UID as attribute for binding`
`The fail one is with use of user@domain` 


What can I do then  ?

**Comment by littlechicks** [comment]
.

**Comment by abentabib** [comment]
.

**Comment by littlechicks** [comment]
So, voilà, the tests failed because we need a active directory server for testing...
Maybe, I can create a Active Directory in Azure Free Trial and test it again ?

**Comment by darrenjennings** [comment]
@kikito instead of patching existing plugin to pluck part of the DN from base_dn, we should instead continue work from https://github.com/Kong/kong/pull/3093 which we are using successfully in the advanced version fo this plugin, which adds read only `config.bind_dn` user who can perform search and use the found user's DN to perform bind.

**Comment by littlechicks** [comment]
@darrenjennings 

Sorry to go back on your comment, but I do not think it's a good solution to recover a password from an AD or any other LDAP Directory.
This is a very bad method. The userpassword attribute should never be retrieved. Otherwise you lose the whole principle of securing a password.

What's more, in active directory it is not possible to retrieve a password attribute without the fUserPwdSupport attribute being FALSE. So it depends on infrastructure.

Similarly in Active Directory a password is encrypted with UniquePwd.
So the pass is hashed, not reversible and it is salted.
And the work done in the PR you describe remains very specific to an OpenLDAP.

To be cross platform, it is better to try and bind with the values ​​provided and either succeed or fail.

In conclusion, ldap plugins kong depends extremly on a DN for binding, so what about an user who is not in the same OU ? It cannot work....

**Comment by littlechicks** [comment]
**Updates**

**First point**
I've installed an Active Directory using Microsoft Azure Free Trial for my tests.
So you cannot not merges my tests before having a working Active Directory with SSL...

**Second Point**
All my tests pass. We can conclude that my code is working with `userPrincipalName`and `CN `(I have to add test for `sAMAccoutName `but it is the same in my case)

**Warning**

- We must use Proxy-Authorization in order to have a successfull authentication. (I don't know why)

- Credentials are passed to the header in the same format `ldap base64(username:password)`
   where username is the real `sAMAccoutName `or the first part of UPN before the `"@"`.

**Comment by darrenjennings** [comment]
> Sorry to go back on your comment, but I do not think it's a good solution to recover a password from an AD or any other LDAP Directory.
> This is a very bad method. The userpassword attribute should never be retrieved. Otherwise you lose the whole principle of securing a password.

Yes the work we ended up using for enterprise did not return password, it uses more traditional method to avoid security issues you mention.

**Comment by littlechicks** [comment]
@darrenjennings 

> > Sorry to go back on your comment, but I do not think it's a good solution to recover a password from an AD or any other LDAP Directory.
> > This is a very bad method. The userpassword attribute should never be retrieved. Otherwise you lose the whole principle of securing a password.
> 
> Yes the work we ended up using for enterprise did not return password, it uses more traditional method to avoid security issues you mention.

Thank you for this info. I am not able to get the code used in advanced plugins so your proposition is to use the PR mentionned as a work base ?
Indeed, having a bind dn who search the DN from an attribute can be useful :)
My solution here is more a workarround for people who want their Active Directory be compatible to the actual plugins, as it easy much simplier than the advanded plugin alternative. What do you think about merging first this workarround and then make a solution based on search function ?

**Comment by omerl13** [comment]
Can you merge this PR please? I need to use the sAMAccountName for authentication (+the OU behavior fix), and right now I can't use Kong due to this issue.

**Comment by nhgiang196** [comment]
I am waiting for this merge/solution so bad. LDAP Auths with sAMAccountName attribute are necessary for my gateway api project

**Review by sosagadinet — COMMENTED** [review]
Hi, there is a typo in:

-- CASE: Attribute is sAMAccountName
if conf.attribute == "sAMAccoutName" then

I merge this code localy and make test against a MS-AD and works.

Thanks for your work.

**Comment by mcen1** [comment]
It still looks like this is an issue, is the code @kikito created going to get merged? Was anyone waiting on this person for anything? Looks like all questions were answered and then the bot marked this as stale. The issue still persists at least in the version of Kong I'm using as of 6 months ago (2.8.1)

**Comment by hanshuebner** [comment]
KAG-1047

**Comment by mcen1** [comment]
I can't do much but suggest this fix is still relevant

**Comment by hbagdi** [comment]
Closing because of lack of activity. We would like to accept this contribution if anyone wishes to pick this back up.

**Comment by Pandede** [comment]
Still waiting for the fix...
